File: code/qa_nonsteadystates.R
#library(tidyverse)
library(readr)

exclude_subs=c('sub-204')

subs = commandArgs(trailingOnly=TRUE)
if (length(subs)==0) {
  subs=c('all')
  #stop("At least one argument must be supplied (input file).n", call.=FALSE)
}

tr <- 0.75
rm_n_trs <- 6 # remove 6 TRs

bids_dir <- '../bids'
#setwd(bids_dir)

# path to confounds folder
fmriprep_dir <- paste0(bids_dir, '/prep/fmriprep')
# list all directories in this folder (subject directories)
all_fmriprep_dirs <- list.dirs(path = fmriprep_dir, full.names = FALSE, recursive = FALSE)

out_dir <- paste0(bids_dir, '/level1/derivatives')

# combine all subjects' confounds tsv files
all_dfs <- c()
for (d in all_fmriprep_dirs) {
  # if this is a subjects folder...
  if (startsWith(d,'sub-')){
    # subject ID
    sub <- d
    if (sub %in% subs | subs == "all") {
      if (sub %in% exclude_subs){
        print(paste('Excluding subject ', sub))
      } else {
        raw_sub_dir <- paste0(bids_dir,'/',sub,'/func')
        events_fnames <- list.files(path = raw_sub_dir, pattern='*tsv', full.names = FALSE)
        # find all tsv files for this subject
        confound_fnames <- list.files(path = paste0(fmriprep_dir,'/',d,'/func'),pattern='*tsv', full.names = TRUE)
        # setup this subject's dataframe
        sub_df=c()
        for (f in confound_fnames){
          base_f <- substr(basename(f), 1,26)
          print(paste('Reading in files starting with:', base_f))
          events_f <- paste0(raw_sub_dir,'/',events_fnames[startsWith(events_fnames, base_f)])
          events_csv <- read.csv(events_f, sep='\t')
          # index for task name in filename
          task_i <- regexpr("task-", f)[1]+5
          # task name
          task <- substring(f,task_i,task_i+5)
          # index for run number in filename
          run_i <- regexpr("run-", f)[1]+4
          # run number
          run <- substr(f,run_i,run_i+1)

          # read in this file
          confound_csv <- read.csv(f, sep = '\t') #, na.strings = 'n/a')
          confound_csv_out <- tail(confound_csv, -rm_n_trs)
          confound_basef <- basename(f)
          out_fname <- paste0(substr(confound_basef, 1, 41), '_rmtr-', as.character(rm_n_trs), substr(confound_basef, 42, 56))
          readr::write_tsv(confound_csv_out, path = paste0(out_dir,'/',sub,'/func/', out_fname))
          print(paste(paste0(out_dir,'/',sub,'/func/', out_fname),"saved"))

          # get number of non-steady state outliers
          nonsteady <- grepl( "non_steady_state_outlier" , names( confound_csv ) )
          # subselect movement parameters
          confound_csv <- as.data.frame(confound_csv[, nonsteady])
          nonsteady_rows <- c()
          nonsteady_time <- 0
          # if nonsteady state columns found...
          if (ncol(confound_csv) > 0){
            warn <- F
            # iterate through columns
            for (i in seq(1,ncol(confound_csv))){
              r <- which(grepl(1, confound_csv[,i]))
              nonsteady_rows <- c(nonsteady_rows, r)
              # check if consecutive
              if (i == 1){
                if (r != 1){
                  print('WARNING: First non-steady state is not first TR!')
                  print(f)
                  warn <- T
                }
              } else {
                r_prev <- nonsteady_rows[i-1]
                if (r != r_prev + 1){
                  print('WARNING: Non-steady states are not consecutive!')
                  print(f)
                  warn <- T
                }
              }
              # get time
              nonsteady_time <- max(nonsteady_rows) * tr
            }
            c_df <- data.frame(nonsteady_rows, task, run, sub)
            # combine this file's data to subject's dataframe
            sub_df <- rbind(sub_df, c_df)
          }
          # recalculate onsets
          events_csv$onset_corrected_by_sub <- events_csv$onset - nonsteady_time
          events_csv$onset_corrected <- events_csv$onset - (rm_n_trs * tr)
          write_tsv(events_csv, path = events_f)
          print(paste(events_f,"saved"))
        }
        # append subject's dataframe to all_dfs
        all_dfs <- rbind(all_dfs, sub_df)
      }
    }
  }
}
fname=paste0(fmriprep_dir,'/nonsteady_outliers.tsv')
write_tsv(all_dfs, path=fname)
print(paste(fname,"saved"))

File: code/qa_nonsteadystates.py
#!/bin/python

import os
import shutil
import sys
from datetime import datetime
from copy import deepcopy
import glob
import pandas as pd
import nibabel as nib
import numpy as np
from scipy.stats import ttest_rel, ttest_ind, ttest_1samp
from funcs import *


print(str(datetime.now()) + ": Begin qa_rm_nonsteadystates.py")

# read in arguments
# subjects
subs = sys.argv[1:]
if len(subs) == 0:
    subs='all'

in_dir = FMRIPREP_DIR
out_dir = DERIVS_DIR

# list all directories in this folder (subject directories)
fnames = os.path.join(FMRIPREP_DIR,
                         "sub-???")
all_fmriprep_dirs = glob.glob(fnames)


# combine all subjects' confounds tsv files
all_dfs = pd.DataFrame(None, columns=['nonsteady_rows', 'task', 'run', 'sub'])
for dir in all_fmriprep_dirs:
    sub = os.path.basename(dir)
    if sub in subs or subs == "all":
      if sub in EXCLUDE_SUBS:
        print(paste('Excluding subject ', sub))
      else:
        raw_sub_dir = os.path.join(BIDS_DIR,sub,'func')
        #events_fnames = glob.glob(os.path.join(raw_sub_dir, '*tsv'))
        # find all tsv files for this subject
        confound_fnames = glob.glob(os.path.join(FMRIPREP_DIR,sub,'func','*tsv'))
        # setup this subject's dataframe
        sub_df = pd.DataFrame(None, columns=['nonsteady_rows', 'task', 'run', 'sub'])
        for f in confound_fnames:
          # get base filename
          confound_basef = f.split('/')[-1]
          t = get_bids_att(confound_basef, 'task')
          r = get_bids_att(confound_basef, 'run')
          #base_f = confound_basef[0:25]
          # read in this file
          confound_df = pd.read_csv(f, delimiter = '\t') #, na.strings = 'n/a')
          confound_df_out = confound_df.iloc[N_TRS_DEL:,:]
          out_fname_components = get_all_bids_atts(confound_basef)
          out_fname_components['rmtr'] = str(N_TRS_DEL)
          if 'extra' in out_fname_components:
              out_fname_components.move_to_end('extra', last=True)
          out_fname = os.path.join(out_dir,
                                    make_bids_str(out_fname_components) +
                                    '.' + confound_basef.split('.')[-1])
          confound_df_out.to_csv(out_fname, sep='\t')
          print(out_fname + " saved")
          #out_fname = paste0(substr(confound_basef, 1, 41), '_rmtr-', as.character(N_TRS_DEL), substr(confound_basef, 42, 56))
          #readr::write_tsv(confound_df_out, path = paste0(out_dir,'/',sub,'/func/', out_fname))
          #print(paste(paste0(out_dir,'/',sub,'/func/', out_fname),"saved"))


          # read in events file
          events_f = glob.glob(os.path.join(raw_sub_dir, sub+'*task-'+t+'*run-'+r+'*tsv'))
          if len(events_f) != 1:
              print("ERROR: "+str(len(events_f))+ " event files found: ")
              print(events_f)
          else:
              events_f = events_f[0]
          print('Reading in file: '+ events_f)
          #events_f = os.path.join(raw_sub_dir,events_fnames+'*')
          events_df = pd.read_csv(events_f, delimiter='\t')

          # # subselect movement parameters
          # confound_df = confound_df.filter(regex='^non_steady_state_outlier',axis=1)
          # nonsteady_rows = []
          # nonsteady_time = 0
          # # if nonsteady state columns found...
          # if confound_df.shape[1] > 0:
          #   warn = False
          #   # iterate through columns
          #   for i in range(confound_df.shape[1]):
          #     r = which(grepl(1, confound_df[,i]))
          #     nonsteady_rows = nonsteady_rows + [r]
          #     # check if consecutive
          #     if (i == 1){
          #       if (r != 1){
          #         print('WARNING: First non-steady state is not first TR!')
          #         print(f)
          #         warn = T
          #       }
          #     } else {
          #       r_prev = nonsteady_rows[i-1]
          #       if (r != r_prev + 1){
          #         print('WARNING: Non-steady states are not consecutive!')
          #         print(f)
          #         warn = T
          #       }
          #     }
          #     # get time
          #     nonsteady_time = max(nonsteady_rows) * TR
          #   }
          #   c_df = data.frame(nonsteady_rows, task, run, sub)
          #   # combine this file's data to subject's dataframe
          #   sub_df = rbind(sub_df, c_df)
          # }
          # # recalculate onsets
          # events_df['onset_corrected_by_sub'] = events_df['onset'] - nonsteady_time

          events_df['onset_corrected'] = events_df['onset'] - (N_TRS_DEL * TR)
          events_df.to_csv(events_f, sep='\t')
          print(events_f + " saved")

        # # append subject's dataframe to all_dfs
        # all_dfs = pd.concat([all_dfs, sub_df])
        
# fname=os.path.join(FMRIPREP_DIR,'nonsteady_outliers.tsv')
# write_tsv(all_dfs, path=fname)
# print(paste(fname,"saved"))

print("Done: qa_nonsteadystates.py")
